File: LeetCode/Simulation/JS/Leetcode1640-check-array-formation-through-concatenation.js
var canFormArray = function(arr, pieces) {
    let mp = new Map();
    for(let i = 0;i < pieces.length;i++) {
        mp.set(pieces[i][0],i);
    }
    for(let i = 0;i < arr.length;) {
        if(!mp.has(arr[i])) return false;
        let t = pieces[mp.get(arr[i])];
        let len = t.length
        for(let j = 0;j < len;j++) {
          if(arr[i] != t[j]) return false;
          i++;
        }
    }
    return true;
};

File: LeetCode/Simulation/Golang/Leetcode1640-check-array-formation-through-concatenation.go
package sbgo

func canFormArray(arr []int, pieces [][]int) bool {
  mp := map[int]int{}
  for i,p := range pieces {
    mp[p[0]] = i
  }
  i,n := 0,len(arr)
  for i < n {
    if _,ok := mp[arr[i]];!ok {
      return false
    }
    t := pieces[mp[arr[i]]]
    size := len(t)
    for j := 0;j < size;j++ {
      if arr[i] != t[j] {
        return false;
      }
      i++
    }
  }
  return true
}
